src/mmw/js/src/router.js

```diff
@@ -73,21 +73,26 @@ var AppRouter = PatchedRouter.extend({
             });
         }
 
+        var self = this;
         this._sequence = this._sequence.then(function() {
             var result = prepare.apply(null, args);
             // Assume result is a promise if an object is returned.
             if (_.isObject(result)) {
+                self._previousRouteName = routeName;
                 return result.then(function() {
                     cb.apply(null, args);
                 });
-            // Only execute the route callback if prepare
-            // did not return false.
             } else if (result !== false) {
+                // Only execute the route callback if prepare
+                // did not return false.
+                self._previousRouteName = routeName;
                 cb.apply(null, args);
+            } else {
+                // Don't execute the route function
+                // and cancel the route transition
+                return false;
             }
         });
-
-        this._previousRouteName = routeName;
```

Comment:
This actually creates subtle queuing problems. If any route is navigated to, before the current route "prepare" method has completed (for example, if there is an animated transition), the sequence will be queued out of order. This could easily happen if a user switches between 2 routes quickly.

The problem isn't obvious because none of our controllers actually return promises (sadly). However, I don't think there's an urgency to fix this until the problem manifests itself.